bindings/rascal/models/krr.py

```diff
@@ -199,11 +232,78 @@ def get_representation_calculator(self):
         return self.kernel._rep
 
 
+# TODO(max, felix) I think this belongs in utils; it's not KRR-specific
+def get_grad_strides(frames):
+    """Get strides for total-energy/gradient kernels of the given structures
+
+    Parameters:
+        frames  List of structures each indicating the number of atoms
+
+    Returns: (1) the number of structures, (2) the number of gradient entries
+    (== 3 * the total number of atoms), and (3) strides for assigning the
+    gradient entries for each structure
+    """
+    Nstructures = len(frames)
+    Ngrad_stride = [0]
+    Ngrads = 0
+    for frame in frames:
+        n_at = len(frame)
+        Ngrad_stride.append(n_at * 3)
+        Ngrads += n_at * 3
+    Ngrad_stride = np.cumsum(Ngrad_stride) + Nstructures
+    return Nstructures, Ngrads, Ngrad_stride
+
+
+def compute_kernel_single(i_frame, frame, representation, X_sparse, kernel):
+    """Compute kernel of the (new) structure against the sparse points
+
+    Parameters:
+        i_frame     frame index (ignored???)
+        frame       New structure to compute kernel for
+        representation
+                    RepresentationCalculator to use for the structures
+        X_sparse    Sparse points to compute kernels against
+        kernel      Kernel object to use
+
+    Return both the kernel and the gradient of the kernel
+    """
+    feat = representation.transform([frame])
+    en_row = kernel(feat, X_sparse)
+    grad_rows = kernel(feat, X_sparse, grad=(True, False))
+    return en_row, grad_rows
+
+
+def compute_KNM(frames, X_sparse, kernel, soap):
+    """Compute kernel of the (new) structure against the sparse points
+
+    Parameters:
+        frame       New structure to compute kernel for
+        representation
+                    RepresentationCalculator to use for the structures
+        X_sparse    Sparse points to compute kernels against
+        kernel      Kernel object to use
+
+    Return the kernel stacked with the gradient of the kernel
+    """
+    Nstructures, Ngrads, Ngrad_stride = get_grad_strides(frames)
+    KNM = np.zeros((Nstructures + Ngrads, X_sparse.size()))
+    pbar = tqdm(frames, desc="compute KNM", leave=False)
+    for i_frame, frame in enumerate(frames):
+        en_row, grad_rows = compute_kernel_single(
+            i_frame, frame, soap, X_sparse, kernel
+        )
+        KNM[Ngrad_stride[i_frame] : Ngrad_stride[i_frame + 1]] = grad_rows
+        KNM[i_frame] = en_row
+        pbar.update()
+    pbar.close()
+    return KNM
+
+
 def train_gap_model(
     kernel,
-    managers,
+    frames,
     KNM_,
-    X_pseudo,
+    X_sparse,
```

Comment:
Mostly unrelated to this PR, but I don't really like the sparse terminology: this is not really sparse, but rather low-rank. Even if the historical literature on GAP is using sparse doesn't mean we have to use it as well =D. 

What do you feel about changing the name to **pseudo points** or **support points** all over librascal?

Reply:
First off, I hate the name "pseudo-points". "Support" or "basis" points is much nicer (and more accurate), but "sparse points" isn't *wrong* (sparse GPs are a thing, see Rasmussen & Williams). If you really want to launch a campaign to change this all across librascal, then by all means, open an issue and let's debate it there.

Reply:
#328